### PR TITLE
New Tuya Serial Based Devices support

### PR DESCRIPTION
**Pull request by thirug010**
Updated the Code to support most of the new Tuya Serial devices like 4Gang Serail switch, 1 Dimmer + 1 relay, 3 Gang

## Description: Updated the Code to set and read the power status along with the device from MCU serial packets, added option to enable or disable the Slider / Dimmer control , added option to set the no of devices (relay / dimmer) Refer [Post](https://github.com/arendst/Sonoff-Tasmota/issues/5545#issuecomment-493599301)

**Related issue (if applicable):** fixes #<Sonoff-Tasmota issue #5545>

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched (Also remember to update _changelog.ino_ file)
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works.
  - [x] The code change pass travis tests. **Your PR cannot be merged unless tests pass**
  - [x] I accept the [CLA](https://github.com/arendst/Sonoff-Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

### CODE REVIEW

**Comment by ascillato** [comment]
Hi,

Please, in order to merge this PR, leave platformio.ini and my_user_config.h without changes. In this PR those files contains your own credentials.

Thanks.

**Comment by thirug010** [comment]
Done - Please check

**Comment by arendst** [comment]
Checked:

1) The use of the pow function increments the code with some kbytes. pow should not be used and be replaced by Tasmota provided FastPrecisePow.
2) There must be a better way to get a binary value from a decimal without using pow function.
3) The use of XdrvMailbox.notused as a global variable should not be used that way. Find another way to get the device number.
4) You might expect that the default setting of bool tuya_show_dimmer = 0. Default disabling the dimmer slide for all non-tuya devices is a wrong option for legacy reasons.

So pls re-chk and verify code size.

**Comment by joba-1** [comment]
Regarding 2)
a) it is not dec, but int -> unintuitive name
b) i see only power of 2 -> use operator << , e.g. pow(2,i) becomes 1 << i

**Comment by thirug010** [comment]
@andrethomas , @joba-1 
Thank you for the input I will check and update you asap 

Thanks 
Thiru

**Comment by andrethomas** [comment]
@thirug010 I think you mean @arendst but I agree with him so I guess there's no harm :)

**Comment by thirug010** [comment]
oh sorry,  wrong tagging I meant @arendst

**Comment by thirug010** [comment]
Updated the code as per the feedback, Please check and let me know 

Changes:

* Updated xdrv_01_webserver.ino to show dimmer slider2 by default for Tuya Dimmer Module

 * Updated xdrv_16_tuyadimmer.ino remove the use of XdrvMailbox and removed the pow functions now the size for 6.5.0 build is 510kb

 * updated sonoff.ino to hold current device id

**Comment by joba-1** [comment]
Hi @thirug010 ,
thanks for your work! Hope you dont mind me nitpicking at your code - I did not review everything, but since the pow() got me interested I am hooked now :)

If you want, you can replace your TuyaGetPower() with this:

```
// p is bitfield (0..7) of power capability for devices d (1..8)
// returns true if device is power capable 

bool TuyaGetPower(uint8_t p, uint8_t d) {
  return (--d < 8)       // true, if device between 1 and 8
    && (p & (1 << d));   // and its power bit is set
}

```

I tested it yields the same result for all possible inputs, and I guess with a bit less code and execution time. If you want to verify: compile below e.g. with gcc -Wall -o testtuya testtuya.cpp):

```
#include <stdio.h>
#include <stdint.h>

// p is bitfield (0..7) of power capability for devices d (1..8)
// returns true if device is power capable 

bool TuyaGetPower(uint8_t p, uint8_t d) {
  return (--d < 8)       // true, if device between 1 and 8
    && (p & (1 << d));   // and its power bit is set
}


uint8_t TuyaGetPower_loop(uint8_t p, uint8_t d) {
  uint8_t g = 0;
  for(int i = 7 ; i >= 0 ; i--) {
    g = p >> i;
    //AddLog_P2(LOG_LEVEL_DEBUG, PSTR("TYA: --SetDevicePower3-- for Device = %d,  Power=%d"),i, k&1);
    if (i== d-1)
    {
       return g&1;
    }
  }
  return 0; // added to be warning free and check all d, not just 1-8
}


int main() {
  uint8_t p=0;
  while(true) {
    printf("\np=%u: d=", p);
    uint8_t d=0;
    while(true) {
      printf("%u,", d);
      if( TuyaGetPower(p, d) != TuyaGetPower_loop(p, d) ) {
        printf("\nerror\n");
        return 1;
      }
      if( ++d == 0 ) break;
    }
    if( ++p == 0 ) break;
  }
  printf("\nok\n");
  return 0;
}

```

**Comment by thirug010** [comment]
@joba-1 , Thanks for feedback, I was missing the point, it is just bitRead from power bit fields, 
I have updated the code, Please check and let me know.

Thanks 
Thiru

**Comment by joba-1** [comment]
hi, looks good to me.
-- 
iphone so i typo

File: sonoff/xdrv_16_tuyadimmer.ino [hunk 1]
```diff
@@ -194,11 +195,11 @@ void TuyaPacketProcess(void)
     case TUYA_CMD_STATE:
       if (tuya_buffer[5] == 5) {  // on/off packet
 
-        AddLog_P2(LOG_LEVEL_DEBUG, PSTR("TYA: RX - %s State"),tuya_buffer[10]?"On":"Off");
-
-        if((power || Settings.light_dimmer > 0) && (power != tuya_buffer[10])) {
+        AddLog_P2(LOG_LEVEL_DEBUG, PSTR("TYA: RX Device -%d --> %s State"),tuya_buffer[6],tuya_buffer[10]?"On":"Off");
+        ExecuteCommandPower(tuya_buffer[6], tuya_buffer[10], SRC_SWITCH);  // send SRC_SWITCH? to use as flag to prevent loop from inbound states from faceplate interaction
```
Comment (chaosmaster) [inline sonoff/xdrv_16_tuyadimmer.ino]: I think this may need to be handled differently.
Without the if it will likely break support for the "Touch Dimmer"

**Comment by thirug010** [comment]
@chaosmaster, Thanks for the Feedback, The code was working with the touch dimmers, I tested the code with theses dimmers ([dimmer1](https://www.amazon.com/Upgraded-Separable-Replaceable-Function-Dressing/dp/B07CKNWGN4/ref=sr_1_fkmrnull_3_sspa?keywords=oittm+smart+dimmer&qid=1558381584&s=hi&sr=1-3-fkmrnull-spons&psc=1), [dimmer2](https://www.amazon.com/gp/product/B07DHGQDQR/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&th=1). 
I compared the results with 6.5.0, I see it is sending Mqtt power status more than once. 

It is fixed now, tested and compare the results with 6.5.0 version refer attached debug logs
[Touch Timmer logs.zip](https://github.com/arendst/Sonoff-Tasmota/files/3199797/Touch.Timmer.logs.zip)

**Comment by chaosmaster** [comment]
I am talking about this one https://www.aliexpress.com/item/WiFi-LED-Dimmer-Switch-220V-110V-Dimming-Panel-Switch-Connected-To-Alexa-Google-Home-Voice-Control/32859257784.html
It looks similar to the second one you linked to, not sure if it's the same though.
This one has always been a bit picky when it comes to handling the feedback from the MCU.
I'll see if I can get around to test your changes on this device tomorrow.

**Comment by thirug010** [comment]
I have the similar device 4 different vendors all have the same serial codes, moreover all the tuya devices needs for follow the same protocols as below, 
image
55aa00 07 0005 01 01 0001 01 0f - ON
55aa00 07 0005 01 01 0001 00 0e - OFF
if it was working with 6.5.0 version then it has the device id 1 so you should be good, Please let me know if you see any issues 

Thanks
Thiru

**Comment by MikeSaint** [comment]
Hi @thirug010 

I'm getting some random dimmer data coming through on my 4 gang touch switch which is making switch 1 turn on by itself if any of the other switches are on.

Is it possible to exclude or ignore any dimmer data if the switch type is 'non-dimmer'?

This is the log data from console. When i press the touch switch to turn on relay 4. Normal behaviour.

09:33:46 TYA: RX Packet: "55aa01070005040100010113"
09:33:46 TYA: RX Device -4 --> On State
09:33:46 SRC: Switch
09:33:46 TYA: Tuya Init Settings.param(2) --> 7
09:33:46 TYA: Tuya Init Settings.param(6) --> 4
09:33:46 RSL: RESULT = {"POWER4":"ON"}
09:33:46 RSL: POWER4 = ON

This is what happens anywhere from 0 - 10 seconds later.

09:33:46 TYA: RX Packet: "55aa0107000807020004000000001c"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa0107000808020004000000001d"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa0107000809020004000000001e"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa010700080a020004000000001f"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa0107000865020004000000007a"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa0107000866020004000000007b"
09:33:46 TYA: RX Dim State=0
09:33:46 TYA: RX Packet: "55aa0107000867020004000000007c"
09:33:46 TYA: RX Dim State=0
**09:33:46 TYA: RX Packet: "55aa0107000868020004000008bb40"
09:33:46 TYA: RX Dim State=187
09:33:46 TYA: Send CMND_DIMMER_STR=Dimmer 73
09:33:46 SRC: Switch
09:33:46 RSL: Received Topic /Dimmer, Data Size 2, Data 73
09:33:46 RSL: Group 0, Index 1, Command DIMMER, Data 73
09:33:46 RSL: RESULT = {"POWER1":"ON","Dimmer":73}**

This is when it’s receiving Dimmer data so is automatically turning on switch 1.

It does the same thing if any of switch 2, 3 or 4 are on.

**Comment by thirug010** [comment]
@MikeSaint,
The Data you shared does not look like dimmer packet, The dimmer packet should have the dimmer id mostly 2, 3 or 1 in some case.  but your serail data has device 9, 65, 66 ... 0a

55aa00070005 01 010001000e		 --> Switch  (device id 1)
55aa00070008 02 020004000000b6cc  --> Dimmer  (device Id 2)
55aa00070005 01 010001000e		 --> Switch  (device id 2)
55aa01070005 03 0100010112		 --> Switch  (device Id 3)
55aa01070005 04 0100010113		 --> Switch  (device Id 4)

based on the Serial data either more than 4 devices configured in the MCU or MCU has issue of sending status for random packets with random device id
55aa01070008 09 020004000000001e
55aa01070008 07 020004000000001c
55aa01070008 08 020004000000001d
55aa01070008 09 020004000000001e
55aa01070008 0a 020004000000001f
55aa01070008 65 020004000000007a
55aa01070008 66 020004000000007b
55aa01070008 67 020004000000007c
55aa01070008 68 020004000008bb40

I will update the code to skip dimmer packet if the device is confgiured as non-dimmer to prevent issue with unknown packets.

Thanks 
Thiru

**Comment by thirug010** [comment]
@MikeSaint , Please Take the Latest from this branch, Check and let me know if the issue is resolved

**Comment by MikeSaint** [comment]
Unfortunately it didn't go well. The firmware compiled fine but gave an 'Upload buffer miscompare' error when I did the OTA upgrade (lack of space probably). This unit refuses to do go into flashing mode on USB. I loaded the minimal firmware OTA and then the full sonoff.bin OTA which loaded successfully. It connected to my wi-fi first time around but the web interface page started dropping in and out repeatedly. After a few reboots it won't connect to Wi-Fi at all. Without the ability to connect to it over serial, I think it's bricked.

**Comment by thirug010** [comment]
The changes were verified with following devices before merge to make sure we did not have any issues,
1. Tuya Dimmer already supported in 6.5.0 ([dimmer1](https://www.amazon.com/Upgraded-Separable-Replaceable-Function-Dressing/dp/B07CKNWGN4/ref=sr_1_fkmrnull_3_sspa?keywords=oittm+smart+dimmer&qid=1558381584&s=hi&sr=1-3-fkmrnull-spons&psc=1), [dimmer2](https://www.amazon.com/gp/product/B07DHGQDQR/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&th=1))
2. New Tuya Dimmer add support in the dev branch ( [New Dimmer](https://www.amazon.com/Smart-Fan-Switch-Compatible-Assistant/dp/B07QRGJGNJ/ref=pd_cp_201_1?pd_rd_w=SUHwI&pf_rd_p=ef4dc990-a9ca-4945-ae0b-f8d549198ed6&pf_rd_r=FBQFKRVD4HBGXRDCN4T9&pd_rd_r=b245bda0-7d63-11e9-b0f3-5fc5494b71c2&pd_rd_wg=DCFPE&pd_rd_i=B07QRGJGNJ&psc=1&refRID=FBQFKRVD4HBGXRDCN4T9))
3. Simple Tuya switch non dimmer ([Tuya-Switch](https://www.amazon.com/Smart-Wifi-Light-Switch-Wall/dp/B07JYR4P5F/ref=pd_sbs_469_5/131-1197570-4835234?_encoding=UTF8&pd_rd_i=B07JYR4P5F&pd_rd_r=e6bde723-7d63-11e9-ba98-0ff29dc49ff2&pd_rd_w=zugHR&pd_rd_wg=lnoEV&pf_rd_p=588939de-d3f8-42f1-a3d8-d556eae5797d&pf_rd_r=26XQPQHFA5XMCABZPEEK&psc=1&refRID=26XQPQHFA5XMCABZPEEK))
4. Sonoff - Non Tuya device just to make sure no issues with sonoff based devices (s[onoff Switch](https://www.amazon.com/dp/B077DXNBY3/ref=twister_B077F7G9Y6?_encoding=UTF8&psc=1))

Having said that I have re-test with couple devices with 6.5.0 to 6.5.0.11 by uploading it couple of times. I am not running in to the issues you are reporting. 

Couple of things to verify,
1. Please check the My_userconfig for the WIFI settings ( as per you comments, it was connected for 1st time, so it might not be the case, but worth to verify it)
2. Disconnect and re-connect the power and check is there is any sonoff-xxx or smartlife-xxxx wifi ssid 

Thanks
Thiru

**Comment by MikeSaint** [comment]
Hi. Agree. I don’t think it’s the code at all. This particular touch switch model is proving problematic. I’ll try another and recompile the code with my wifi settings rather than factory defaults.

**Comment by thirug010** [comment]
OK sure, please let me know.

**Comment by n3PH1lim** [comment]
Hi i purchased this device: https://www.amazon.de/gp/product/B07M7Z249P/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1

This is one of this switches esp->tuya Serial device.

Tasmota flash worked but i need this fix for correct operation. 

When it is in the development branch i can test it.

Will the be an configuration template like the tuya dimmer?

Thx

**Comment by thirug010** [comment]
@n3PH1lim,
 We cannot tell if the device use serial packet or use GPIO, 

1. set weblog to 4 to check by turn on/off of the switch  and see console serial data is send then it is dimmer template 
2. if no serial data then it uses gpio you have to find it configuration by checking the gpio's is relay or LED or Button 

thanks 
thiru

**Comment by n3PH1lim** [comment]
To get anything I must set the device to Tuya Dimmer.
I can only change 1 relay with this configuration but I get the following logs:

1st Button:
18:39:38 TYA: TX Packet: "55aa0006000800020004000000192c"
18:39:38 RSL: stat/sonoff/RESULT = {"POWER":"ON"}
18:39:38 RSL: stat/sonoff/POWER = ON
18:39:38 CFG: Saved to flash at FA, Count 74, Bytes 3584
18:39:45 TYA: TX Packet: "55aa00000000ff"
18:39:45 TYA: RX Packet: "55aa000000010101"


2nd Button:
8:40:09 TYA: TX Packet: "55aa0006000800020004000000192c"
18:40:09 RSL: stat/sonoff/RESULT = {"POWER":"ON"}
18:40:09 RSL: stat/sonoff/POWER = ON
18:40:09 TYA: RX Packet: "55aa00070005030100010111"
18:40:09 TYA: RX - On State

3rd Button:
18:40:51 TYA: RX Packet: "55aa0007000501010001010f"
18:40:51 TYA: RX - On State
18:40:51 TYA: RX Packet: "55aa00070005020100010110"
18:40:51 TYA: RX - On State
18:40:51 TYA: RX Packet: "55aa00070005030100010111"
18:40:51 TYA: RX - On State
18:40:52 WIF: Checking connection...
18:40:52 WIF: Connected
18:41:02 TYA: TX Packet: "55aa00000000ff"
18:41:02 TYA: RX Packet: "55aa000000010101"

**Comment by thirug010** [comment]
@n3PH1lim,

This looks like the correct Tuya packets for Switch 1,2,3. Others are random or unknown packets  you can 
- 18:40:51 TYA: RX Packet: "55aa0007000501010001010f"
- 18:40:51 TYA: RX Packet: "55aa00070005020100010110"
- 18:40:51 TYA: RX Packet: "55aa00070005030100010111"
Use options SetOption63 as 1 and SetOption38 as 3 and restart device. it should work, Please check and let me know 

Thanks 
Thiru

**Comment by n3PH1lim** [comment]
I have tested it and it works :-D

**Comment by thirug010** [comment]
Thank You!

**Comment by andrethomas** [comment]
Don't mind me... I'm just trying to see how these issues can be resolved as usually, they would end up having to be re-based as the PR includes files which we don't necessarily want to commit/merge.

**Comment by andrethomas** [comment]
There, I removed the troubling files :)

**Comment by thirug010** [comment]
@andrethomas Thank You !

**Pull request merged**